_posts/2023-04-21-tensorflow-mnist-classification.adoc

```diff
@@ -0,0 +1,405 @@
+In the future, Debezium potential can be increased further, e.g. by extending CDC not only to databases.
+
+So, this is the theory.
+Now let's explore how it works in reality.
+In this blog post we will look at how to stream data into TensorFlow.
+Based on the interest from the community, this may result in a series of blog posts where we try to explore possible integrations with other ML libraries and frameworks.
+
+== Debezium and TensorFlow integration
+
+TensorFlow is one of the most popular machine learning frameworks.
+It provides a comprehensive platform for building, training, and deploying machine learning models across a wide range of applications.
+
+To keep things simple, we will implement a model for recognition of handwritten digits, which is more or less the Hello World equivalent in the neural networks field.
+The ultimate goal of this demo is to load MNIST data samples from Postgres using Debezium, pass it to our model implemented in Tensorflow for training and use this trained model for real time classification of images which are continuously stored in the database.
+
+The complete pipeline is depicted in the diagram bellow:
+
+++++
+<div class="imageblock centered-image">
+    <img src="/assets/images/2023-04-21-tensorflow-mnist-classification/pipeline.png" class="responsive-image" alt="Debezium TensorFlow integration pipeline">
+</div>
+++++
+
+All the code mentioned later in this blog is available as a Debezium example in http://TBD[Debezium example repository].
+
+=== The data sample
+
+We will use http://yann.lecun.com/exdb/mnist/[MNIST data sample].
+The training sample contains 60,000 images with handwritten digits from 0 to 9 and the same amount of labels with corresponding digits.
+Test sample contains 1,000 images.
+The samples are available as gzip binaries.
+As we assume a use case where the data of interest are in the database, we need to load the data into the database first.
+
+We need to generate two SQL files, one for the train data set, `mnist_train.sql`, and one for test data sample, `mnist_test.sql`.
+Each file would contain SQL commands for creating table with two columns: `pixels` column of type `BYTEA`, which would contain raw image bytes and `labels` column of type `SMALLINT`, which would contain digit corresponding to the image in given table row.
+Rest of the file would contain commands for populating the table.
+Image bytes can be decoded as a HEX string.
+
+As we will show how to leverage Debezium also for data streaming later in this post, we will initially load into the database only training data set.
+SQL file with training data will be used directly by Postgres container - when it starts, it will load this data into the training table.
+The test data SQL file will be used later on.
+However, the preparation of the data is the same for training as well as test samples and we can prepare both of them in one go.
```

Comment:
```suggestion
As we will show how to leverage Debezium for data streaming later in this post, we will initially load the training data set into the database.
The SQL file with training data will be used directly by the Postgres container - when it starts, it will load this data into the training table.
We will use the test data SQL file later.
However, the preparation of the data is the same for training as well as test samples, and we can prepare both of them in one go.
```